In [1]:
import pandas as pd
import os, sys



file_dir = os.path.dirname(os.path.abspath("__file__"))
parent_dir = os.path.dirname(file_dir)
newPath = os.path.join(parent_dir, 'data/db_historic/rt_leavetimes_2016_I_DB.txt')

# read leavetimes 2016 .txt file into csv leave16 (just first 100)
leave16 = pd.read_csv(newPath, delimiter=';', chunksize=100)


# just put 10 in a dataframe for the moment...
leave16chunk=leave16.get_chunk(10)
leave16chunk

datasource        dayofservice   tripid  progrnumber  stoppointid  \
0         DB  31-JAN-16 00:00:00  2811044           63         3390   
1         DB  31-JAN-16 00:00:00  2820852            1         7158   
2         DB  31-JAN-16 00:00:00  2820852            5         7017   
3         DB  31-JAN-16 00:00:00  2820852           10         1893   
4         DB  31-JAN-16 00:00:00  2820853           16         1648   
5         DB  31-JAN-16 00:00:00  2820853           21         1805   
6         DB  31-JAN-16 00:00:00  2820853           36         1854   
7         DB  31-JAN-16 00:00:00  2820853           55         1899   
8         DB  31-JAN-16 00:00:00  2820853           60         7240   
9         DB  31-JAN-16 00:00:00  2820854            7         7030   

   plannedtime_arr  plannedtime_dep  actualtime_arr  actualtime_dep  \
0            41820            41820           42920           42920   
1            69300            69300           69320           69320   
2            69498            69498           69532           69532   
3            69711            69711           69858           69858   
4            76822            76822           76636           76657   
5            76989            76989           76826           76826   
6            77792            77792           77700           77715   
7            78676            78676           78609           78609   
8            78798            78798           78844           78844   
9            80378            80378           80352           80352   

   vehicleid  passengers  passengersin  passengersout  distance  suppressed  \
0    1000897         NaN           NaN            NaN       NaN         NaN   
1    1000903         NaN           NaN            NaN       NaN         NaN   
2    1000903         NaN           NaN            NaN       NaN         NaN   
3    1000903         NaN           NaN            NaN       NaN         NaN   
4    1000903         NaN           NaN            NaN       NaN         NaN   
5    1000903         NaN           NaN            NaN       NaN         NaN   
6    1000903         NaN           NaN            NaN       NaN         NaN   
7    1000903         NaN           NaN            NaN       NaN         NaN   
8    1000903         NaN           NaN            NaN       NaN         NaN   
9    1000903         NaN           NaN            NaN       NaN         NaN   

   justificationid          lastupdate  note  
0              NaN  12-APR-16 05:30:20   NaN  
1              NaN  12-APR-16 05:30:20   NaN  
2              NaN  12-APR-16 05:30:20   NaN  
3              NaN  12-APR-16 05:30:20   NaN  
4              NaN  12-APR-16 05:30:20   NaN  
5              NaN  12-APR-16 05:30:20   NaN  
6              NaN  12-APR-16 05:30:20   NaN  
7              NaN  12-APR-16 05:30:20   NaN  
8              NaN  12-APR-16 05:30:20   NaN  
9              NaN  12-APR-16 05:30:20   NaN

In [2]:
leave16chunk.dtypes

datasource          object
dayofservice        object
tripid               int64
progrnumber          int64
stoppointid          int64
plannedtime_arr      int64
plannedtime_dep      int64
actualtime_arr       int64
actualtime_dep       int64
vehicleid            int64
passengers         float64
passengersin       float64
passengersout      float64
distance           float64
suppressed         float64
justificationid    float64
lastupdate          object
note               float64
dtype: object

In [3]:
leave16chunk

datasource        dayofservice   tripid  progrnumber  stoppointid  \
0         DB  31-JAN-16 00:00:00  2811044           63         3390   
1         DB  31-JAN-16 00:00:00  2820852            1         7158   
2         DB  31-JAN-16 00:00:00  2820852            5         7017   
3         DB  31-JAN-16 00:00:00  2820852           10         1893   
4         DB  31-JAN-16 00:00:00  2820853           16         1648   
5         DB  31-JAN-16 00:00:00  2820853           21         1805   
6         DB  31-JAN-16 00:00:00  2820853           36         1854   
7         DB  31-JAN-16 00:00:00  2820853           55         1899   
8         DB  31-JAN-16 00:00:00  2820853           60         7240   
9         DB  31-JAN-16 00:00:00  2820854            7         7030   

   plannedtime_arr  plannedtime_dep  actualtime_arr  actualtime_dep  \
0            41820            41820           42920           42920   
1            69300            69300           69320           69320   
2            69498            69498           69532           69532   
3            69711            69711           69858           69858   
4            76822            76822           76636           76657   
5            76989            76989           76826           76826   
6            77792            77792           77700           77715   
7            78676            78676           78609           78609   
8            78798            78798           78844           78844   
9            80378            80378           80352           80352   

   vehicleid  passengers  passengersin  passengersout  distance  suppressed  \
0    1000897         NaN           NaN            NaN       NaN         NaN   
1    1000903         NaN           NaN            NaN       NaN         NaN   
2    1000903         NaN           NaN            NaN       NaN         NaN   
3    1000903         NaN           NaN            NaN       NaN         NaN   
4    1000903         NaN           NaN            NaN       NaN         NaN   
5    1000903         NaN           NaN            NaN       NaN         NaN   
6    1000903         NaN           NaN            NaN       NaN         NaN   
7    1000903         NaN           NaN            NaN       NaN         NaN   
8    1000903         NaN           NaN            NaN       NaN         NaN   
9    1000903         NaN           NaN            NaN       NaN         NaN   

   justificationid          lastupdate  note  
0              NaN  12-APR-16 05:30:20   NaN  
1              NaN  12-APR-16 05:30:20   NaN  
2              NaN  12-APR-16 05:30:20   NaN  
3              NaN  12-APR-16 05:30:20   NaN  
4              NaN  12-APR-16 05:30:20   NaN  
5              NaN  12-APR-16 05:30:20   NaN  
6              NaN  12-APR-16 05:30:20   NaN  
7              NaN  12-APR-16 05:30:20   NaN  
8              NaN  12-APR-16 05:30:20   NaN  
9              NaN  12-APR-16 05:30:20   NaN

In [4]:
# use chunk size

chunk_size = 1000000
count=1
suppressedIsOneCount=0

# load the big file in smaller chunks
for leave_chunk in pd.read_csv(newPath, delimiter=';', chunksize=chunk_size):
    print("--------------------")
    print("Chunk number: ", count)
    print()
    print("datasource has value null in this chunk:", leave_chunk['datasource'].isnull().sum())
    print("dayofservice has value null in this chunk:", leave_chunk['dayofservice'].isnull().sum())
    print("tripid has value null in this chunk:", leave_chunk['tripid'].isnull().sum())
    print("progrnumber has value null in this chunk:", leave_chunk['progrnumber'].isnull().sum())
    print("stoppointid has value null in this chunk:", leave_chunk['stoppointid'].isnull().sum())
    print("plannedtime_arr has value null in this chunk:", leave_chunk['plannedtime_arr'].isnull().sum())
    print("plannedtime_dep has value null in this chunk:", leave_chunk['plannedtime_dep'].isnull().sum())
    print("actualtime_arr has value null in this chunk:", leave_chunk['actualtime_arr'].isnull().sum())
    print("actualtime_dep has value null in this chunk:", leave_chunk['actualtime_dep'].isnull().sum())
    print("vehicleid has value null in this chunk:", leave_chunk['vehicleid'].isnull().sum())
    print("passengers has value null in this chunk:", leave_chunk['passengers'].isnull().sum())
    print("passengersin has value null in this chunk:", leave_chunk['passengersin'].isnull().sum())
    print("passengersout has value null in this chunk:", leave_chunk['passengersout'].isnull().sum())
    print("distance has value null in this chunk:", leave_chunk['distance'].isnull().sum())
    print("suppressed has value null in this chunk:", leave_chunk['suppressed'].isnull().sum())
    print("justificationid has value null in this chunk:", leave_chunk['justificationid'].isnull().sum())
    print("lastupdate has value null in this chunk:", leave_chunk['lastupdate'].isnull().sum())
    print("note has value null in this chunk:", leave_chunk['note'].isnull().sum())
    print()
    # check for duplicate rows
    print('Duplicate rows:', leave_chunk.duplicated()[leave_chunk.duplicated() == True].shape[0])
    # Check for duplicate columns
    print('Duplicate columns:',leave_chunk.columns.size - leave_chunk.columns.unique().size)
    print()
    print("FINISHED A CHUNK")
    print()
    print()
    suppressed_is_1 = leave_chunk['suppressed']==1  
    #print("suppressed: ", leave_chunk.head())
    is_suppressed = leave_chunk[suppressed_is_1]
    print("Number of rows in this chunk where suppressed column has value 1: ", is_suppressed.shape[0])
    print()
    #leave_chunk.to_csv('leave16_prelim.csv', index=True, header=False, mode='a')   
    suppressedIsOneCount=suppressedIsOneCount+is_suppressed.shape[0]
    count=count+1
    #if count==4:
       # break
        
        
print()
print("Total no. of suppressed = 1 values: ", suppressedIsOneCount)
   
    
print()
print("CELL FINISHED")


--------------------
Chunk number:  1

datasource has value null in this chunk: 0
dayofservice has value null in this chunk: 0
tripid has value null in this chunk: 0
progrnumber has value null in this chunk: 0
stoppointid has value null in this chunk: 0
plannedtime_arr has value null in this chunk: 0
plannedtime_dep has value null in this chunk: 0
actualtime_arr has value null in this chunk: 0
actualtime_dep has value null in this chunk: 0
vehicleid has value null in this chunk: 0
passengers has value null in this chunk: 1000000
passengersin has value null in this chunk: 1000000
passengersout has value null in this chunk: 1000000
distance has value null in this chunk: 1000000
suppressed has value null in this chunk: 961971
justificationid has value null in this chunk: 961971
lastupdate has value null in this chunk: 0
note has value null in this chunk: 1000000

Duplicate rows: 0
Duplicate columns: 0

FINISHED A CHUNK


Number of rows in this chunk where suppressed column has value 1:  3

stoppointid has value null in this chunk: 0
plannedtime_arr has value null in this chunk: 0
plannedtime_dep has value null in this chunk: 0
actualtime_arr has value null in this chunk: 0
actualtime_dep has value null in this chunk: 0
vehicleid has value null in this chunk: 0
passengers has value null in this chunk: 1000000
passengersin has value null in this chunk: 1000000
passengersout has value null in this chunk: 1000000
distance has value null in this chunk: 1000000
suppressed has value null in this chunk: 962352
justificationid has value null in this chunk: 962352
lastupdate has value null in this chunk: 0
note has value null in this chunk: 1000000

Duplicate rows: 0
Duplicate columns: 0

FINISHED A CHUNK


Number of rows in this chunk where suppressed column has value 1:  37648

--------------------
Chunk number:  10

datasource has value null in this chunk: 0
dayofservice has value null in this chunk: 0
tripid has value null in this chunk: 0
progrnumber has value null in this ch

justificationid has value null in this chunk: 963665
lastupdate has value null in this chunk: 0
note has value null in this chunk: 1000000

Duplicate rows: 0
Duplicate columns: 0

FINISHED A CHUNK


Number of rows in this chunk where suppressed column has value 1:  36335

--------------------
Chunk number:  18

datasource has value null in this chunk: 0
dayofservice has value null in this chunk: 0
tripid has value null in this chunk: 0
progrnumber has value null in this chunk: 0
stoppointid has value null in this chunk: 0
plannedtime_arr has value null in this chunk: 0
plannedtime_dep has value null in this chunk: 0
actualtime_arr has value null in this chunk: 0
actualtime_dep has value null in this chunk: 0
vehicleid has value null in this chunk: 0
passengers has value null in this chunk: 1000000
passengersin has value null in this chunk: 1000000
passengersout has value null in this chunk: 1000000
distance has value null in this chunk: 1000000
suppressed has value null in this chunk: 

Duplicate rows: 0
Duplicate columns: 0

FINISHED A CHUNK


Number of rows in this chunk where suppressed column has value 1:  27705

--------------------
Chunk number:  26

datasource has value null in this chunk: 0
dayofservice has value null in this chunk: 0
tripid has value null in this chunk: 0
progrnumber has value null in this chunk: 0
stoppointid has value null in this chunk: 0
plannedtime_arr has value null in this chunk: 0
plannedtime_dep has value null in this chunk: 0
actualtime_arr has value null in this chunk: 0
actualtime_dep has value null in this chunk: 0
vehicleid has value null in this chunk: 0
passengers has value null in this chunk: 1000000
passengersin has value null in this chunk: 1000000
passengersout has value null in this chunk: 1000000
distance has value null in this chunk: 1000000
suppressed has value null in this chunk: 962942
justificationid has value null in this chunk: 962942
lastupdate has value null in this chunk: 0
note has value null in this chunk: 10

stoppointid has value null in this chunk: 0
plannedtime_arr has value null in this chunk: 0
plannedtime_dep has value null in this chunk: 0
actualtime_arr has value null in this chunk: 0
actualtime_dep has value null in this chunk: 0
vehicleid has value null in this chunk: 0
passengers has value null in this chunk: 1000000
passengersin has value null in this chunk: 1000000
passengersout has value null in this chunk: 1000000
distance has value null in this chunk: 1000000
suppressed has value null in this chunk: 967686
justificationid has value null in this chunk: 967686
lastupdate has value null in this chunk: 0
note has value null in this chunk: 1000000

Duplicate rows: 0
Duplicate columns: 0

FINISHED A CHUNK


Number of rows in this chunk where suppressed column has value 1:  32314

--------------------
Chunk number:  35

datasource has value null in this chunk: 0
dayofservice has value null in this chunk: 0
tripid has value null in this chunk: 0
progrnumber has value null in this ch

suppressed has value null in this chunk: 967272
justificationid has value null in this chunk: 967272
lastupdate has value null in this chunk: 0
note has value null in this chunk: 1000000

Duplicate rows: 0
Duplicate columns: 0

FINISHED A CHUNK


Number of rows in this chunk where suppressed column has value 1:  32728

--------------------
Chunk number:  43

datasource has value null in this chunk: 0
dayofservice has value null in this chunk: 0
tripid has value null in this chunk: 0
progrnumber has value null in this chunk: 0
stoppointid has value null in this chunk: 0
plannedtime_arr has value null in this chunk: 0
plannedtime_dep has value null in this chunk: 0
actualtime_arr has value null in this chunk: 0
actualtime_dep has value null in this chunk: 0
vehicleid has value null in this chunk: 0
passengers has value null in this chunk: 1000000
passengersin has value null in this chunk: 1000000
passengersout has value null in this chunk: 1000000
distance has value null in this chunk: 1

Duplicate rows: 0
Duplicate columns: 0

FINISHED A CHUNK


Number of rows in this chunk where suppressed column has value 1:  33647

--------------------
Chunk number:  51

datasource has value null in this chunk: 0
dayofservice has value null in this chunk: 0
tripid has value null in this chunk: 0
progrnumber has value null in this chunk: 0
stoppointid has value null in this chunk: 0
plannedtime_arr has value null in this chunk: 0
plannedtime_dep has value null in this chunk: 0
actualtime_arr has value null in this chunk: 0
actualtime_dep has value null in this chunk: 0
vehicleid has value null in this chunk: 0
passengers has value null in this chunk: 1000000
passengersin has value null in this chunk: 1000000
passengersout has value null in this chunk: 1000000
distance has value null in this chunk: 1000000
suppressed has value null in this chunk: 962563
justificationid has value null in this chunk: 962563
lastupdate has value null in this chunk: 0
note has value null in this chunk: 10

plannedtime_arr has value null in this chunk: 0
plannedtime_dep has value null in this chunk: 0
actualtime_arr has value null in this chunk: 0
actualtime_dep has value null in this chunk: 0
vehicleid has value null in this chunk: 0
passengers has value null in this chunk: 923405
passengersin has value null in this chunk: 923405
passengersout has value null in this chunk: 923405
distance has value null in this chunk: 923405
suppressed has value null in this chunk: 889348
justificationid has value null in this chunk: 889348
lastupdate has value null in this chunk: 0
note has value null in this chunk: 923405

Duplicate rows: 0
Duplicate columns: 0

FINISHED A CHUNK


Number of rows in this chunk where suppressed column has value 1:  34057


Total no. of suppressed = 1 values:  2080086

CELL FINISHED


CONCLUSIONS:
- There are NO duplicate rows or columns in whole of leave2016 txt file
- Passengers, Passengersin and Passengersout are all null in this file. So the three columns are candidates to be dropped... we'll just check it's the same in the leave2017 txt file as well.
- ALL 'distance' columns are null/NaN, meaning it's a strong candidate to be dropped, assuming all the distance columns in leave2017 txt file are also null/NaN
- ALL 'note' columns are null/NaN, meaning it's a strong candidate to be dropped, assuming all the note columns in leave2017 txt file are also null/NaN
- no. of suppressed=1 values are: 2,080,086. These are to be dropped in final cleanse
- When the 'to_CSV' line is not commented out, this notebook creates a csv with 58923405 lines in it.
- although there are some entries in 'justificationid', as the feature records 'fault codes' to do with why a journey was suppressed - this is not relevant to developing our model, so we will be dropping this column.
- 'lastupdate' is not relevant to our model either, as it does not relate to what day a journey took place, only when the data was updated. So we will drop this column completely also.
- We also have previously discoverd that the feature datasource has a cardinality of one: ie, it only has one value, DB, for the entire dataset. Therefore we will drop this feature.
- The Vehicleid feature will also be dropped as it is not relevant to developing our model.